app/javascript/controllers/message_scroller_controller.js

```diff
@@ -0,0 +1,55 @@
+window.lastMessageControllerInstance = null
+window.wasScrolledToBottom = false
+
+import { Controller } from "@hotwired/stimulus"
+import debounce from "utils/debounce"
+
+export default class extends Controller {
+  scrollableTarget = null
+
+  static values = {
+    scrollDown: { type: Boolean, default: false },
+    instant: { type: Boolean, default: false },
+    onlyScrollDownIfWasBottom: { type: Boolean, default: false }
+  }
+
+  connect() {
+    if (window.lastMessageControllerInstance) {
+      window.lastMessageControllerInstance.disconnect()
+      window.lastMessageControllerInstance = this
+    }
+    this.scrollableTarget = document.getElementById('right-content')
+    window.addEventListener('resize', this.debouncedScrollDownIfWasBottom) // resizing browser & composer size changes
+
+    this.considerScroll()
+  }
+
+  disconnect() {
+    window.removeEventListener('resize', this.debouncedScrollDownIfWasBottom)
+  }
+
+  considerScroll() {
+    if (this.scrollDownValue)
+      this.debouncedScrollDown()
+    else if (this.onlyScrollDownIfWasBottomValue)
+      this.scrollDownIfWasBottom()
+  }
+
+  debouncedScrollDownIfWasBottom = debounce(() => this.scrollDownIfWasBottom(), 50, true)
+  scrollDownIfWasBottom() {
+    if (window.wasScrolledToBottom) this.debouncedScrollDown(false)
+  }
+
+  debouncedScrollDown = debounce((instant) => this.scrollDown(instant), 50, true)
+  scrollDown(instant) {
+    let instantScroll = instant ?? this.instantValue
+    setTimeout(() => {
+      window.wasScrolledToBottom = true // even if we don't get the full way, it was the intention
```

Comment:
should this flag be set outside of the timeout?

Reply:
This was a good catch. You had a couple good flags on logic related to scroll down, and notably, I wasn't totally happy with the scroll down logic. It still was jumpy at times and I still did not have an easy enough time scrolling **up** the page while a conversation was streaming in. I was really hoping that one of these things you pointed out was the root cause.

I fixed them, but unfortunately the issues didn't go away. :) But it led me down the right rabbit hole and I think I now have a root cause fix. I'm about to tee up a new PR.